File: webhook/scripts/export-concept-graph.ts
/**
 * scripts/export-concept-graph.ts
 *
 * Regenerates the small JSON the 3D concept-graph viewer consumes, reading
 * from the MIGRATED Firestore library instead of a session temp file.
 *
 * This replaces scripts/export_concept_graph.py as the source of truth for
 * app/public/full-concept-graph.json. That script projected down from
 * graph_data.json, which lived only in a Claude session's scratchpad; this one
 * reads conceptLibrary, which is durable production data. The OUTPUT SHAPE IS
 * DELIBERATELY IDENTICAL so app/public/full-graph-viewer.html needs no change:
 *
 *     nodes:    {id, name, subject, level, hasLesson, hasSim}
 *     edges:    {source, target, relation, weight}
 *     subjects: {id, title, color, nodeCount, withLesson, withSim}
 *     counts:   {nodes, edges, withLesson, withSim, subjects}
 *
 * One honest difference from the Python version, worth knowing when the two
 * outputs are compared: the Python script set hasSim from the mere PRESENCE of
 * a sim reference (759 concepts). Firestore only holds sims whose built HTML
 * actually existed on disk at migration time, so this reports 722. The smaller
 * number is the true one: 37 of those references point at an extract that
 * never built into a playable file, and advertising a sim that cannot be shown
 * is worse than admitting the gap.
 *
 * Also writes app/public/full-concept-graph-counts.json, a few hundred bytes
 * carrying just the counts and subject rollup, so pages that only need honest
 * totals ("722 of 4118 concepts have a sim") do not download the 1.5 MB graph.
 *
 * Idempotent: nodes and edges are sorted, so the same library always produces
 * the same bytes apart from the generatedAt stamp.
 *
 * Usage (needs FIREBASE_SERVICE_ACCOUNT in the env):
 *   node --env-file=.env.local -r ts-node/register/transpile-only scripts/export-concept-graph.ts
 */
import * as fs from 'fs'
import * as path from 'path'
import { createHash } from 'crypto'
import { db } from '../lib/firebase'
import { CONCEPT_LIBRARY } from '../lib/conceptLibrary'

const OUT_GRAPH = path.resolve(__dirname, '../../app/public/full-concept-graph.json')
const OUT_COUNTS = path.resolve(__dirname, '../../app/public/full-concept-graph-counts.json')

/** Same hand-tuned, hue-spaced palette the Python exporter used, pinned per
 * subject id so adding a subject never recolors an existing one. */
const SUBJECT_COLORS: Record<string, string> = {
  'act-math': '#c4f547',
  'algebra-1': '#8ee06a',
  biology: '#43d98b',
  blockchain: '#f0a93c',
  calculus: '#38d6c4',
  chemistry: '#45c2f0',
  circuits: '#5aa8ff',
  'computer-science': '#8b8cf7',
  networking: '#a97cf0',
  psychology: '#c96ee8',
  'quantum-computing': '#ef6fc6',
  'statistics-course': '#f56a92',
  'us-history': '#f5794f',
}

/** Stable, repeatable color for a subject added after that palette was
 * written. Hue from a hash of the id, saturation/lightness fixed in the same
 * band as the curated colors so it never looks alien. */
function fallbackColor(subjectId: string): string {
  const d = createHash('sha256').update(subjectId).digest()
  const hue = ((d[0] << 8) | d[1]) / 65535
  const [r, g, b] = hslToRgb(hue, 0.72, 0.62)
  return '#' + [r, g, b].map((v) => Math.round(v * 255).toString(16).padStart(2, '0')).join('')
}

function hslToRgb(h: number, s: number, l: number): [number, number, number] {
  if (s === 0) return [l, l, l]
  const q = l < 0.5 ? l * (1 + s) : l + s - l * s
  const p = 2 * l - q
  const hue2rgb = (t: number) => {
    if (t < 0) t += 1
    if (t > 1) t -= 1
    if (t < 1 / 6) return p + (q - p) * 6 * t
    if (t < 1 / 2) return q
    if (t < 2 / 3) return p + (q - p) * (2 / 3 - t) * 6
    return p
  }
  return [hue2rgb(h + 1 / 3), hue2rgb(h), hue2rgb(h - 1 / 3)]
}

function pretty(id: string): string {
  return id.replace(/[-_]/g, ' ').replace(/\b\w/g, (c) => c.toUpperCase())
}

interface OutNode { id: string; name: string; subject: string; level: string; hasLesson: boolean; hasSim: boolean }
interface OutEdge { source: string; target: string; relation: string; weight: number }

async function main() {
  console.log(`reading ${CONCEPT_LIBRARY} from Firestore...`)
  const snap = await db.collection(CONCEPT_LIBRARY).get()
  if (snap.empty) {
    console.error(`ERROR: ${CONCEPT_LIBRARY} is empty. Run scripts/migrate-concept-library.ts --write first.`)
    process.exit(1)
  }

  const nodes: OutNode[] = []
  const subjectTitles = new Map<string, string>()
  const edgeSeen = new Set<string>()
  const edges: OutEdge[] = []
  const ids = new Set<string>()

  interface LibDoc {
    name?: string
    subject?: string
    subjectTitle?: string
    level?: string
    hasLesson?: boolean
    hasSim?: boolean
    prereqs?: string[]
    crossSubject?: string[]
  }

  const docs = snap.docs.map((d) => ({ id: d.id, data: d.data() as LibDoc }))
  for (const { id, data } of docs) {
    ids.add(id)
    const subject = data.subject || 'unknown'
    if (data.subjectTitle) subjectTitles.set(subject, data.subjectTitle)
    nodes.push({
      id,
      name: data.name || id,
      subject,
      level: data.level || 'core',
      hasLesson: data.hasLesson === true,
      hasSim: data.hasSim === true,
    })
  }

  // Edges are reconstructed from the denormalized arrays. `prereqs` on X holds
  // the concepts that come BEFORE X, so the edge is prereq -> X, which is the
  // same direction the source graph's {from, to} carried. crossSubject is
  // stored on both endpoints, so it is de-duplicated by sorted pair.
  let dangling = 0
  for (const { id, data } of docs) {
    for (const from of data.prereqs ?? []) {
      if (!ids.has(from)) { dangling++; continue }
      const key = `${from}|${id}|prerequisite`
      if (edgeSeen.has(key)) continue
      edgeSeen.add(key)
      edges.push({ source: from, target: id, relation: 'prerequisite', weight: 1 })
    }
    for (const other of data.crossSubject ?? []) {
      if (!ids.has(other)) { dangling++; continue }
      const [a, b] = [id, other].sort()
      const key = `${a}|${b}|cross_subject`
      if (edgeSeen.has(key)) continue
      edgeSeen.add(key)
      edges.push({ source: a, target: b, relation: 'cross_subject', weight: 0.5 })
    }
  }

  nodes.sort((a, b) => (a.subject === b.subject ? (a.id < b.id ? -1 : 1) : a.subject < b.subject ? -1 : 1))
  edges.sort((a, b) =>
    a.source !== b.source ? (a.source < b.source ? -1 : 1)
      : a.target !== b.target ? (a.target < b.target ? -1 : 1)
        : a.relation < b.relation ? -1 : 1,
  )

  const bySubject = new Map<string, { id: string; title: string; color: string; nodeCount: number; withLesson: number; withSim: number }>()
  for (const n of nodes) {
    let s = bySubject.get(n.subject)
    if (!s) {
      s = {
        id: n.subject,
        title: subjectTitles.get(n.subject) ?? pretty(n.subject),
        color: SUBJECT_COLORS[n.subject] ?? fallbackColor(n.subject),
        nodeCount: 0,
        withLesson: 0,
        withSim: 0,
      }
      bySubject.set(n.subject, s)
    }
    s.nodeCount++
    if (n.hasLesson) s.withLesson++
    if (n.hasSim) s.withSim++
  }
  const subjects = [...bySubject.values()].sort((a, b) => (a.id < b.id ? -1 : 1))

  const counts = {
    nodes: nodes.length,
    edges: edges.length,
    withLesson: nodes.filter((n) => n.hasLesson).length,
    withSim: nodes.filter((n) => n.hasSim).length,
    subjects: subjects.length,
  }
  const generatedAt = new Date().toISOString().replace(/\.\d{3}Z$/, 'Z')

  fs.mkdirSync(path.dirname(OUT_GRAPH), { recursive: true })
  fs.writeFileSync(
    OUT_GRAPH,
    JSON.stringify({ generatedAt, source: `firestore:${CONCEPT_LIBRARY}`, counts, subjects, nodes, edges }),
  )
  fs.writeFileSync(
    OUT_COUNTS,
    JSON.stringify({ generatedAt, source: `firestore:${CONCEPT_LIBRARY}`, counts, subjects }),
  )

  console.log(`wrote   ${OUT_GRAPH} (${(fs.statSync(OUT_GRAPH).size / 1024).toFixed(0)} KB)`)
  console.log(`wrote   ${OUT_COUNTS} (${fs.statSync(OUT_COUNTS).size} bytes)`)
  console.log(`nodes   ${counts.nodes} (${counts.withLesson} with a lesson, ${counts.withSim} with a sim)`)
  console.log(`edges   ${counts.edges} across ${counts.subjects} subjects`)
  if (dangling) console.log(`note    skipped ${dangling} edge endpoints pointing at unknown concepts`)
  const uncolored = subjects.filter((s) => !SUBJECT_COLORS[s.id]).map((s) => s.id)
  if (uncolored.length) console.log(`note    new subjects on a generated color: ${uncolored.join(', ')}`)
}

main().then(() => process.exit(0)).catch((e) => {
  console.error(e)
  process.exit(1)
})

File: scripts/export_concept_graph.py
#!/usr/bin/env python3
"""Export the live content graph into the small JSON the 3D viewer consumes.

Source of truth is the content pipeline's graph_data.json (nodes with real
lesson objects and sims, prerequisite + cross_subject edges). That file is
big (tens of MB) and full of chapter bodies the viewer never needs, so this
script projects it down to the minimum the browser needs to draw and
highlight the graph:

    nodes: {id, name, subject, level, hasLesson, hasSim}
    edges: {source, target, relation, weight}

Re-run this after every content-merge round. It is idempotent: nodes and
edges are sorted, so the same input always produces the same graph payload
(only the generatedAt stamp moves), and everything it reports (node counts,
subject list, colors) is derived from the file at run time, never hardcoded.

Usage:
    python3 scripts/export_concept_graph.py
    python3 scripts/export_concept_graph.py --source /path/to/graph_data.json
    python3 scripts/export_concept_graph.py --out app/public/full-concept-graph.json

The source path can also come from the MINDCRAFT_GRAPH_DATA env var.
"""

from __future__ import annotations

import argparse
import colorsys
import hashlib
import json
import os
import sys
from datetime import datetime, timezone
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]

DEFAULT_SOURCE = Path(
    "/private/tmp/claude-501/-Users-akoirala/ecdb2be0-20e4-4b62-a08a-9acc53a4fbc5"
    "/scratchpad/ios-preview/graph_data.json"
)
DEFAULT_OUT = ROOT / "app" / "public" / "full-concept-graph.json"

# Hand-tuned, hue-spaced palette. Subjects are alphabetical and their hues walk
# the color wheel, so neighbours in the legend are never confusable, and every
# color reads clearly against the viewer's near-black background. Pinned per
# subject id (not per index) so adding a subject never recolors existing ones.
SUBJECT_COLORS = {
    "act-math": "#c4f547",
    "algebra-1": "#8ee06a",
    "biology": "#43d98b",
    "blockchain": "#f0a93c",
    "calculus": "#38d6c4",
    "chemistry": "#45c2f0",
    "circuits": "#5aa8ff",
    "computer-science": "#8b8cf7",
    "networking": "#a97cf0",
    "psychology": "#c96ee8",
    "quantum-computing": "#ef6fc6",
    "statistics-course": "#f56a92",
    "us-history": "#f5794f",
}


def fallback_color(subject_id: str) -> str:
    """Stable color for a subject that appears after this palette was written.

    Hue comes from a hash of the id, so a brand new subject gets a usable,
    repeatable color with zero code changes. Saturation and lightness are
    fixed in the same band as the curated colors so it does not look alien.
    """
    digest = hashlib.sha256(subject_id.encode("utf-8")).digest()
    hue = (digest[0] << 8 | digest[1]) / 65535.0
    r, g, b = colorsys.hls_to_rgb(hue, 0.62, 0.72)
    return "#{:02x}{:02x}{:02x}".format(int(r * 255), int(g * 255), int(b * 255))


def pretty(subject_id: str) -> str:
    return subject_id.replace("-", " ").replace("_", " ").title()


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--source",
        type=Path,
        default=Path(os.environ.get("MINDCRAFT_GRAPH_DATA", DEFAULT_SOURCE)),
        help="graph_data.json produced by the content pipeline",
    )
    parser.add_argument("--out", type=Path, default=DEFAULT_OUT, help="viewer JSON to write")
    args = parser.parse_args()

    source: Path = args.source
    if not source.exists():
        print(f"ERROR: source graph not found: {source}", file=sys.stderr)
        print("Pass --source or set MINDCRAFT_GRAPH_DATA.", file=sys.stderr)
        return 1

    raw = json.loads(source.read_text())
    raw_nodes = raw.get("nodes") or []
    raw_edges = raw.get("edges") or []
    subject_titles = {s["id"]: s.get("title") or pretty(s["id"]) for s in (raw.get("subjects") or [])}

    nodes = []
    seen: set[str] = set()
    duplicates = 0
    for n in raw_nodes:
        node_id = n.get("id")
        if not node_id:
            continue
        if node_id in seen:
            duplicates += 1
            continue
        seen.add(node_id)
        lesson = n.get("lesson") or None
        sims = n.get("sims") or []
        nodes.append(
            {
                "id": node_id,
                "name": n.get("name") or node_id,
                "subject": n.get("subject") or "unknown",
                "level": n.get("level") or "core",
                "hasLesson": bool(lesson and (lesson.get("body") or lesson.get("summary"))),
                "hasSim": bool(sims),
            }
        )

    # Sorted for stable diffs across re-runs. Ids come in two shapes, bare
    # ("circles_geometry", act-math only) and subject-prefixed
    # ("biology::phylogenetics"), and both are carried through untouched:
    # the viewer and the search resolver key off whatever id lands here.
    nodes.sort(key=lambda n: (n["subject"], n["id"]))

    edges = []
    dangling = 0
    edge_seen: set[tuple[str, str, str]] = set()
    for e in raw_edges:
        src, dst = e.get("from"), e.get("to")
        if not src or not dst:
            continue
        if src not in seen or dst not in seen:
            dangling += 1
            continue
        relation = e.get("relation") or "prerequisite"
        key = (src, dst, relation)
        if key in edge_seen:
            continue
        edge_seen.add(key)
        edges.append(
            {
                "source": src,
                "target": dst,
                "relation": relation,
                "weight": round(float(e.get("weight") or 0.0), 3),
            }
        )
    edges.sort(key=lambda e: (e["source"], e["target"], e["relation"]))

    # Per-subject rollup, computed from what is actually in the file today.
    by_subject: dict[str, dict] = {}
    for n in nodes:
        s = by_subject.setdefault(
            n["subject"],
            {
                "id": n["subject"],
                "title": subject_titles.get(n["subject"], pretty(n["subject"])),
                "color": SUBJECT_COLORS.get(n["subject"]) or fallback_color(n["subject"]),
                "nodeCount": 0,
                "withLesson": 0,
                "withSim": 0,
            },
        )
        s["nodeCount"] += 1
        s["withLesson"] += 1 if n["hasLesson"] else 0
        s["withSim"] += 1 if n["hasSim"] else 0
    subjects = sorted(by_subject.values(), key=lambda s: s["id"])

    payload = {
        "generatedAt": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
        "source": str(source),
        "counts": {
            "nodes": len(nodes),
            "edges": len(edges),
            "withLesson": sum(1 for n in nodes if n["hasLesson"]),
            "withSim": sum(1 for n in nodes if n["hasSim"]),
            "subjects": len(subjects),
        },
        "subjects": subjects,
        "nodes": nodes,
        "edges": edges,
    }

    args.out.parent.mkdir(parents=True, exist_ok=True)
    args.out.write_text(json.dumps(payload, separators=(",", ":")))

    c = payload["counts"]
    print(f"source  {source}")
    print(f"wrote   {args.out}  ({args.out.stat().st_size / 1024:.0f} KB)")
    print(
        f"nodes   {c['nodes']}  ({c['withLesson']} with a lesson, {c['withSim']} with a sim)"
    )
    print(f"edges   {c['edges']}  across {c['subjects']} subjects")
    if duplicates:
        print(f"note    skipped {duplicates} duplicate node ids")
    if dangling:
        print(f"note    skipped {dangling} edges pointing at unknown nodes")
    missing = [s["id"] for s in subjects if s["id"] not in SUBJECT_COLORS]
    if missing:
        print(f"note    new subjects on a generated color: {', '.join(missing)}")
        print("        add them to SUBJECT_COLORS in this script for a hand-picked hue")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
